core/crypto/src/vrf.rs

```diff
@@ -157,182 +149,30 @@ impl SecretKey {
-                let mut buf = [0; $l];
-                if bs58::decode(value).into(&mut buf[..]) == Ok($l) {
-                    return Self::try_from(&buf).or(Err(Error));
-                }
-                Err(Error)
-            }
-        }
-
-        impl TryFrom<String> for $ty {
-            type Error = Error;
-            fn try_from(value: String) -> Result<Self, Error> {
-                Self::try_from(value.as_str())
-            }
-        }
-
-        impl Into<[u8; $l]> for $ty {
-            fn into(self) -> [u8; $l] {
-                *self.as_ref()
-            }
-        }
+    ($ty:ident, $l:literal, $bytes:expr, $what:literal) => {
+        eq!($ty, |a, b| a.0 == b.0);
+        common_conversions_fixed!($ty, 32, $bytes, $what);
 
-        impl Into<[u8; $l]> for &$ty {
-            fn into(self) -> [u8; $l] {
-                *self.as_ref()
-            }
-        }
-
-        impl Into<String> for $ty {
-            fn into(self) -> String {
-                bs58::encode(self).into_string()
-            }
-        }
-
-        impl Into<String> for &$ty {
-            fn into(self) -> String {
-                bs58::encode(self).into_string()
-            }
-        }
-
-        impl Debug for $ty {
-            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-                f.write_str(to_str!(self))
-            }
-        }
-
-        impl Display for $ty {
-            fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-                f.write_str(to_str!(self))
-            }
-        }
-
-        impl<'de> Deserialize<'de> for $ty {
-            fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
-                let s = <&str as Deserialize<'de>>::deserialize(deserializer)?;
-                Self::try_from(s).map_err(|_| {
-                    D::Error::invalid_value(Unexpected::Str(s), &concat!("a valid ", $what))
-                })
-            }
-        }
-
-        impl Serialize for $ty {
-            fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
-                serializer.serialize_str(to_str!(self))
-            }
-        }
-    };
-}
-
-macro_rules! traits_k {
-    ($ty:ident, $l:literal) => {
         impl TryFrom<&[u8; $l]> for $ty {
-            type Error = Error;
-            fn try_from(value: &[u8; $l]) -> Result<Self, Error> {
-                Self::from_bytes(value).ok_or(Error)
-            }
-        }
-    };
-}
-
-macro_rules! traits_v {
-    ($ty:ident, $l:literal) => {
-        impl AsMut<[u8; $l]> for $ty {
-            fn as_mut(&mut self) -> &mut [u8; $l] {
-                &mut self.0
-            }
-        }
-
-        impl AsMut<[u8]> for $ty {
-            fn as_mut(&mut self) -> &mut [u8] {
-                &mut self.0[..]
-            }
-        }
-
-        impl From<&[u8; $l]> for $ty {
-            fn from(value: &[u8; $l]) -> Self {
-                Self(*value)
+            type Error = ();
```

Comment:
Why is error type `()`? If you don't need an error type why implement `TryFrom`?

Reply:
I need a unit type for error. That is, there is a possibility of error, but there is no additional information to be provided with an error.

Reply:
Why is there no additional information?

Reply:
Because a value is either valid or not. There is no need for additional information.